_source/_posts/2019-09-19-nodejs-typescript.md

```diff
@@ -0,0 +1,637 @@
+```typescript
+import { Controller, Get } from '@nestjs/common';
+import { ApiModelProperty, ApiResponse } from '@nestjs/swagger';
+
+export class BlogPost {
+  @ApiModelProperty() // used to generate Swagger documentation that `BlogPost` model contains id of type number
+  id: number;
+
+  @ApiModelProperty()
+  authorId: string;
+
+  @ApiModelProperty()
+  title: string;
+
+  @ApiModelProperty()
+  content: string;
+}
+
+export const blogPosts = [{
+  id: 1,
+  authorId: 'xxx',
+  title: 'Build a NodeJS App with Typescript',
+  content: 'Whats wrong with Javascript?'
+}, {
+  id: 2,
+  authorId: 'yyy',
+  title: 'Don\'t build a NodeJS App with Typescript',
+  content: 'Whats wrong with Typescript?'
+}];
+
+@Controller('blog-posts')
+export default class BlogPostController {
+  @Get() // registers a `blog-posts` GET method on the API
+  @ApiResponse({ type: BlogPost, status: 200, isArray: true }) // for Swagger documentation: API returns an array of BlogPost models
+  findAll(): Array<BlogPost> {
+    return blogPosts;
+  }
+}
+```
+
+Refresh browser page `http://localhost:3000`, and execute `GET blog-posts` in the API console.
+
+## Create the Authentication module
+
+Create a directory for authentication module: `src/auth.module`.
+
+To help offload the burden of authentication, you will use Okta for authentication, authorization, and user account management.
+
+## Create a forever-free Okta account and application
+
+[Click here](https://developer.okta.com/signup/) to create an account.
+
+Once you're logged into your Okta dashboard, click **Applications** in the menu and click **Add Application**. From the wizard, choose **Service** and click **Next**.
+
+On the Application Settings screen, name the application (I've named mine "the-blog-backend").
+
+After an application has been successfully created, copy the client ID and client secret into the `.env` file in the root of the project.
+
+The dotenv (.env) file stores your development environment variables. They switch your application's code from development to production or reference accounts on 3rd-party providers.
+
+This file is consumed by the `require('dotenv/config')` statement in your `server.js` file. It copies the values from `.env` file into the `process.env` global variable. Create an empty `.env` file in the root directory of the project.
+
+The application needs an API token to communicate with Okta. In the Okta dashboard, click on **API**, then **Tokens**, and on the new page click**Create Token**. Copy its value to the `.env` file as well.
+
+You'll also put your Okta account's domain in the `.env` file. It's the hostname of the Okta dashboard you're currently using, but without the `-admin` part. It should look something like `dev-xxxxxx.okta.com`.
+
+After writing all those values, your `.env` file should look something like this:
+
+```dotenv
+OKTA_CLIENT_ID={yourClientId}
+OKTA_CLIENT_SECRET={yourClientSecret}
+OKTA_APP_TOKEN={yourAppToken}
+OKTA_DOMAIN={yourDomainUrl}
```

Comment:
Should the domain have https:// in front of it? If so, we should add it here and in the paragraph above.